src/controller/testController.ts

```diff
@@ -258,6 +271,133 @@ export const runTests: (request: TestRunRequest, option: IRunOption) => any = in
+                    projectName: testContext.projectName,
+                    testKind: TestKind.None,
+                    testLevel: TestLevel.Invocation,
+                }, parentItem, []);
+            }
+            if (event.displayName && getLabelWithoutCodicon(currentItem.label) !== event.displayName) {
+                currentItem.description = event.displayName;
+            }
+            switch (event.state) {
+                case TestResultState.Running:
+                    run.started(currentItem);
+                    break;
+                case TestResultState.Passed:
+                    run.passed(currentItem);
+                    break;
+                case TestResultState.Failed:
+                case TestResultState.Errored:
+                    const testMessages: TestMessage[] = [];
+                    if (event.message) {
+                        const markdownTrace: MarkdownString = new MarkdownString();
+                        markdownTrace.supportHtml = true;
+                        markdownTrace.isTrusted = true;
+                        const testMessage: TestMessage = new TestMessage(markdownTrace);
+                        testMessages.push(testMessage);
+                        const lines: string[] = event.message.split(/\r?\n/);
+                        for (const line of lines) {
+                            const location: Location | undefined = processStackTraceLine(line, markdownTrace, currentItem, testContext.projectName);
+                            if (location) {
+                                testMessage.location = location;
+                            }
+                        }
+                    }
+                    run.failed(currentItem, testMessages);
+                    break;
+                case TestResultState.Skipped:
+                    run.skipped(currentItem);
+                    break;
+                default:
+                    break;
+            }
+        });
+
+        testRunner.onDidFinishTestRun((_event: TestFinishEvent) => {
+            return resolve();
+        });
+
+        testRunner.launch(testContext);
+    });
+
+    function findTestClass(projectName: string, testIdentifier: string): TestItem {
+        const projectItem: TestItem | undefined = testController?.items.get(projectName);
+        if (!projectItem) {
+            throw new Error('Failed to get the project test item.');
+        }
+        const idxOfMethodStart: number = testIdentifier.indexOf('#');
+        let classFullyQualifiedName: string;
+        if (idxOfMethodStart > 0) {
+            classFullyQualifiedName = testIdentifier.substring(0, idxOfMethodStart);
+        } else {
+            classFullyQualifiedName = testIdentifier;
+        }
+
+        const idxOfLastDot: number = classFullyQualifiedName.lastIndexOf('.');
+        const packageName: string = idxOfLastDot > 0 ? classFullyQualifiedName.substring(0, idxOfLastDot) : '';
```

Comment:
I see lots of places (e.g. vscode-gradle, test runner extension) to parse the testId to extract the package, class, method of test cases. Can you write two utilities such as `convertIdToTestItem`, `convertTestItemToId` for reuse?

Reply:
Either you can define a new model such as `JavaTestItem` for internal communication between test runner and gradle extension. The testId is only used for exchange between test runner and vscode client.